builtin/common/misc_helpers.lua

```diff
@@ -517,6 +517,66 @@ function table.shuffle(t, from, to, random)
 end
 
 
+function table.find(t, e, i, j)
+    i = i or 1
+    j = j or #t
+    
+    local mt = {__eq = function(e1, e2)
```

Comment:
Probably better to use a local function rather than __eq, to avoid needing to override the operator. You should also fall back to an existing metatable, anyway

ie:

```lua
local function are_equal(a, b)
	if a == b then
		return true
	end
	if type(b) ~= "table" then
		return false
	end

	-- table checking logic here
end
```

Reply:
Maybe it is better in respect of a less operations number. However, there is a more serious problem: how to cope if a searched table has still nested tables? That would lead to an indefinite number of nested loops is what obviously impossible. Also, I dont't know how to compare userdatas since they are C-side objects.

Reply:
> That would lead to an indefinite number of nested loops is what obviously impossible

No, you can use recursion

Reply:
> No, you can use recursion

As far as I know, full table comparisons are difficult in some cases. Here's an example where `t1` and `t2` are both recursively nested empty tables:
```lua
t1 = {}
t1[1] = t1

t2 = {}
t = {}
t2[1] = t
t[1] = t2
```

Reply:
I'm not convinced it's even worth doing `table.find`, tbh. There's already `table.indexof` for `==` matches, all it adds is deep table comparisons

Reply:
table.indexof like returns only first entry whereas my function returns a whole list with entries of the element that is searched out. Also, that first method ignores non-numerical indexes and thus any element can not be found in tables with keys of other types.

Reply:
Also, it can search in a certain range over a table.

Reply:
> Also, it can search in a certain range over a table.

It uses `pairs` for the loop and increments `iter` in each step. The order in which `pairs` iterates over the table is undefined, and `iter` is used to check if the iteration is in this range. Therefore I assume that the range search works incorrectly.

Reply:
Hmm, probably you're right. Anyway it could be avoided by transmitting all values in a list and iterating it.